[IPO] Added attributor for identifying invariant loads

zGoldthorpe opened the pull request:
The attributor conservatively marks pointers whose loads are eligible to be marked as `!invariant.load`.
It does so by identifying:
1. Pointers marked `noalias` and `readonly`
2. Pointers whose underlying objects are all eligible for invariant loads.

The attributor then manifests this attribute at non-atomic non-volatile load instructions.

zGoldthorpe committed:
Added attributor for identifying `!invariant.load`s.

zGoldthorpe commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,248 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+      return ChangeStatus::CHANGED;
+    }
+    if (ANoAlias->isKnownNoAlias())
+      addKnownBits(IS_NOALIAS);
```
Yes, the intention was to return `UNCHANGED` after this. Based on my understanding of `indicateOptimisticFixpoint()` also returning `UNCHANGED`, I figured the same should happen here since the attributor already assumes `IS_NOALIAS` at this point.

(Or maybe I misunderstood the feedback?)

zGoldthorpe committed:
Incorporated feedback

zGoldthorpe commented on llvm/test/Transforms/Attributor/tag-invariant-loads.ll:
```diff
@@ -0,0 +1,220 @@
+define i32 @test_noalias_ptr(ptr noalias %ptr) {
+; CHECK-LABEL: define i32 @test_noalias_ptr(
+; CHECK-SAME: ptr noalias nofree noundef nonnull readonly align 4 captures(none) dereferenceable(4) [[PTR:%.*]]) #[[ATTR0]] {
+; CHECK-NEXT:    [[VAL:%.*]] = load i32, ptr [[PTR]], align 4, !invariant.load [[META0:![0-9]+]]
```
I think you're right. I misunderstood the strength of the assumptions on an `!invariant.load`. I'm not sure if there is a useful way to salvage the attributor at this level of generality. It's probably best to close this PR and narrow the scope, as you suggest.

arsenm replied:
Yes, readonly can be inferred for atomic loads

krzysz00 replied:
Hm, yeah, then a lack of atomic access to a thing is also important

... Right, so the one case is spin-waiting on a flag that some other device (the host, say), will set, which might still _look_ `readonly noalias` but shouldn't be invariant.

... Actually, no, if you're doing that, I think you want to be volatile on top of being atomic? My recollection of pointer semnatics is that "load from a location I never write whose value mysteriously changes" needs to be volatile.

So I think we're clear to mark non-volatile loads invariant if they meet the other conditions, because if you're doing atomic-y things, you'll either have a code path that writes the location or you need to be volatile.

zGoldthorpe replied:
For the record, the attributor *does* add `readonly` to `%flag`

arsenm replied:
readonly doesn't imply the memory doesn't change, it only states that the function does not write through a reference derived from this argument. I don't think the AA property does anything here

zGoldthorpe replied:
I have improved the analysis to detect effectful atomic loads from pointers, which should address these concerns.

zGoldthorpe commented:
I have tried addressing most of the feedback.
However, I haven't done anything regarding atomics, yet.

zGoldthorpe committed:
Added guards for side-effects on loads.

"Side effects" include volatile loads and atomic loads that are at least
monotonic.

krzysz00 reviewed (commented):
This _feels_ reasonable, but I don't fully know the attributer framework well enough to review it properly.

Also, do we want to add some mechanism for declaring that ptr(4) and ptr(6) are all definitionally constant and use it here?

Finally, call for a test
```
define amdgpu_kernel void @test_noalias_buffer_desc(ptr addrspace(1) align 4 noalias %ptr) {
  %buffer.rsrc = call ptr addrspace(7) @llvm.amdgcn.make.buffer.rsrc.p7.p1(ptr addrspace(1) %ptr, i16 0, i32 256, i32 0)
  %val = load i32, ptr addrspace(7) %buffer.rsrc, align 4
  call void @clobber(i32 %val)
  ret void
}
```

zGoldthorpe committed:
Corrected and refactored attributor logic.

zGoldthorpe committed:
Modified checks for unrelated but affected tests

It seems the attributor cleans up more dead instructions.

zGoldthorpe commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,269 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+    if (isCallableCC(F->getCallingConv())) {
+      // effects cannot be tracked outside of function call;
+      // conservatively assume pointer has effectful uses
+      removeAssumedBits(IS_NOEFFECT);
```
The `IS_NOEFFECT` bit checks for more than just communication, and also checks for other side effects involving the pointer (such as being written through).

As for `@test_noalias_buffer_desc`, I'm not sure if there is a nice generalisable way to mark the load as invariant in this case, besides explicitly checking for such intrinsics as `@llvm.amdgcn.make.buffer.rsrc.p7.p1` and forwarding information.

> Also, do we want to add some mechanism for declaring that ptr(4) and ptr(6) are all definitionally constant and use it here?

Is there existing infrastructure that indicates whether address spaces are definitionally constant based on the target architecture?

shiltian commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,301 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+    Changed |= checkNoEffect(A);
+
+    // try to infer invariance from underlying objects
+    const auto *AUO = A.getOrCreateAAFor<AAUnderlyingObjects>(
```
check `nullptr` and its validity

shiltian commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,301 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+        return true;
+
+      if (auto *LI = dyn_cast<LoadInst>(I)) {
+
```
```suggestion
```

shiltian commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,301 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+            getIRPosition(), this, DepClassTy::REQUIRED)) {
+      if (ANoAlias->isKnownNoAlias()) {
+        addKnownBits(IS_NOALIAS);
+        return ChangeStatus::UNCHANGED;
```
Don't we change something here by adding the know bit?

shiltian commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,301 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+      return ChangeStatus::UNCHANGED;
+    }
+
+    if (const auto *Arg = getAssociatedArgument()) {
```
no auto

zGoldthorpe committed:
Incorporated feedback.

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,317 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+  void trackStatistics() const override {}
+
+protected:
+  /// Indicate that invariance necessarily requires the pointer to be noalias.
+  virtual bool requiresNoAlias() const { return false; }
```
Don't understand this, but it probably shouldn't be a virtual

arsenm commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,317 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+    const Function *F = getAssociatedFunction();
+    assert(F && "no associated function for return from call");
+
+    // not much we can say about opaque functions
```
Capitalize comments

arsenm commented on llvm/test/Transforms/Attributor/tag-invariant-loads.ll:
```diff
@@ -0,0 +1,324 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -mtriple=amdgcn-amd-amdhsa -passes=attributor %s -S | FileCheck %s --check-prefix=AMDGCN
```
Probably needs to be in AMDGPU subdirectory or REQUIRES: amdgpu-registered-target

arsenm commented on llvm/test/Transforms/Attributor/tag-invariant-loads.ll:
```diff
@@ -0,0 +1,324 @@
+declare ptr addrspace(1) @get_ptr() #0
+declare noalias ptr addrspace(1) @get_noalias_ptr() #0
+declare noalias ptr addrspace(1) @get_untouched_ptr() #1
+attributes #0 = { nofree norecurse nosync nounwind willreturn }
+attributes #1 = { nofree norecurse nosync nounwind willreturn readonly }
```
Keep these at the end of the file

zGoldthorpe committed:
Incorporated feedback.

krzysz00 commented:
> No, this file is target agnostic so all target related stuff can be here. If anything target related needed, it has to go through TTI hook.

Yeah, that's what I meant. `TTI::isGlobalConstantsAddressSpace(unsigned)` if such a thing doesn't already exist

krzysz00 commented:
Yeah, separate PR, but worth doing as a followup

shiltian commented:
> > Yeah, that's what I meant. `TTI::isGlobalConstantsAddressSpace(unsigned)` if such a thing doesn't already exist
> 
> I don't think such a thing already exists, but I feel like adding such support would make more sense as a separate PR.

Ideally yes, but in practice maybe challenging. I've tried to add stuff about address spaces into some common files several times in the past but got a lot of push back. Almost none of those effort succeeded. The essential question is the definition. For this particular case, I'd imagine the challenge is how to define a "global constant" address space? What is its characteristic?

zGoldthorpe commented:
> Ideally yes, but in practice maybe challenging. I've tried to add stuff about address spaces into some common files several times in the past but got a lot of push back. Almost none of those effort succeeded. The essential question is the definition. For this particular case, I'd imagine the challenge is how to define a "global constant" address space? What is its characteristic?

Maybe a constant addrspace should be analogous to (`external`) `constant` [global variables](https://llvm.org/docs/LangRef.html#globalvars)? E.g., we may assume that the contents of the address space are **never** modified, and `store`s to it are UB.

arsenm commented:
> > Ideally yes, but in practice maybe challenging. I've tried to add stuff about address spaces into some common files several times in the past but got a lot of push back. Almost none of those effort succeeded. The essential question is the definition. For this particular case, I'd imagine the challenge is how to define a "global constant" address space? What is its characteristic?
> 
> Maybe a constant addrspace should be analogous to (`external`) `constant` [global variables](https://llvm.org/docs/LangRef.html#globalvars)? E.g., we may assume that the contents of the address space are **never** modified, and `store`s to it are UB.

No, constant global variable is already a constant. The constant address space is wholly redundant. We should work to eliminate the notion of constant address space we have now. I wouldn't put any effort into trying to infer constantness out of the address space

krzysz00 commented:
If we want to get rid of the constant address space idiom, that's fine by me - we'll just want to get the frontends to tag their loads from constants as invariant

zGoldthorpe commented:
@jdoerfert pinging to request review

krzysz00 commented:
> The main problem is invariant load doesn't do anything on memcpy or other memory intrinsics

That feels ... fixable? (Especially since MachineMemOperand has an `invariant` field)

But that's probably a low-priority eventually sort of thing

shiltian commented on llvm/lib/Transforms/IPO/AttributorAttributes.cpp:
```diff
@@ -12534,6 +12535,318 @@ struct AAIndirectCallInfoCallSite : public AAIndirectCallInfo {
+  bool requiresNoAlias() const {
+    switch (getPositionKind()) {
+    default:
+      return false;
```
Do you need to be conservative for `default` here?

zGoldthorpe replied:
Yeah, I should be more conservative.

zGoldthorpe committed:
Made default `noalias` requirement more conservative.

shiltian reviewed (approved):
LGTM. @krzysz00 WDYT?

krzysz00 reviewed (approved):
I think this is good from what I can tell.

Although one ask - either here in or in a followup PR - can we make sure that this still works with `ptr addrspace(7)`s created by `make.buffer.rsrc.p7.p1`? (Or that stuff like `invariant.start` doesn't break the analysis?)

zGoldthorpe commented:
> I think this is good from what I can tell.
> 
> Although one ask - either here in or in a followup PR - can we make sure that this still works with `ptr addrspace(7)`s created by `make.buffer.rsrc.p7.p1`? (Or that stuff like `invariant.start` doesn't break the analysis?)

I feel like this extension makes more sense in a follow-up PR, since it would probably be nicer to have a more uniform solution to address such intrinsics.

zGoldthorpe committed:
Added support for certain intrinsics.

zGoldthorpe commented:
@jdoerfert pinging for review
